Defibrillator Fixes + Code Cleanup

SpaceDragon00 opened the pull request:
## About The Pull Request

Cleans up some of the defib code by reducing repeated code blocks, renaming certain procs, reducing unnecessary if statements, and adding flexability to certain aspects, like cooldown time. Additionally fixes some issues with the code, like visible messages not showing up for mobs who are wearing the source and shock_touching not using its provided damage values.

I have defined the combat paddles' emp_act change as a "Fix" because the comment appended to the combat variable, as well as the lack of sprites/clarity for emped syndicate defibs, made me feel this was unintentional. However, if this is seen as intentional, I will revert those changes to keep this a purely fix/cleanup-based PR

## Why It's Good For The Game

Cleaner code and fewer bugs

## Changelog
:cl:
fix: Combat defibs no longer enable their safeties when being EMPed
fix: Visible message sends its message to mobs that are also its location
fix: shock_touching (Renamed to shock_pulling) now actually uses the damage value provided
code: Removed useless mob/user input for cooldowncheck
code: Added recharge_time variable to paddles for variable charge times
code: Added can_hurt, do_success, and do_cancel to reduce code duplication
/:cl:

SpaceDragon00 committed:
Initial Improvements + Bugfixes

- Combat defibs can no longer be EMPed to enable their safeties
- Removed useless mob/user input for cooldowncheck
- Added recharge_time variable to paddles for variable charge times
- Added can_hurt, do_success, and do_cancel to reduce code duplication
- shock_touching (Renamed to shock_pulling) now actually uses the damage value provided
- Visible message sends its message to mobs that are also its location

Colovorat commented:
>fix: Combat defibs no longer enable their safeties when being EMPed
So when i did this one, it was balance change that was not merged...
Huh...

SpaceDragon00 commented:
> 
> 
> > fix: Combat defibs no longer enable their safeties when being EMPed
> 
> So when i did this one, it was balance change that was not merged...
> Huh...

I haven't gotten confirmation on whether or not it's intentional yet, that's just my initial feeling. I'll change it back if such is requested

SpaceDragon00 committed:
Merge remote-tracking branch 'upstream/master' into defibfixes

SpaceDragon00 committed:
Merge remote-tracking branch 'upstream/master' into defibfixes

Rohesie commented on code/game/objects/items/defib.dm:
```diff
@@ -468,17 +473,43 @@
 	do_help(H, user)
 
-/obj/item/shockpaddles/proc/shock_touching(dmg, mob/H)
+/// Returns TRUE when the paddles can do harmful actions, and false if they cant (If the safety is off, or if they're combat paddles)
+/obj/item/shockpaddles/proc/can_hurt()
+	if(combat)
+		return TRUE
+	if(req_defib && !defib.safety)
+		return TRUE
+	return FALSE
```
I'm generally not a big fan of `is_thing()` or `can_do_thing()` procs that just check several conditions.
They tend to grow, be redefined for extra exceptions, and then make producing events-oriented code way harder.

My suggestion is to either make `can_hurt` a variable that changes based on the changes of the other variables (through setters), or alternatively just to turn this into a macro, which has lower chances of being expanded into:
`#define SHOCKPADDLES_CAN_HURT (combat || (req_defib && !defib.safety))`

SpaceDragon00 replied:
Good point, I'll turn it into a define

SpaceDragon00 committed:
Merge remote-tracking branch 'upstream/master' into defibfixes

SpaceDragon00 committed:
Removed can_hurt Proc, Substituted with Define

- Removed the can_hurt() proc and replaced it with DEFIB_CAN_HURT

Rohesie commented:
Should be good after the merge conflict is solved.

SpaceDragon00 committed:
Merge remote-tracking branch 'upstream/master' into defibfixes

Rohesie commented on code/game/objects/items/defib.dm:
```diff
@@ -161,7 +162,11 @@
-	if(safety)
+
+	if(combat) // Elite agents do not subscribe to your notion of "Safety"
+		visible_message("<span class='notice'>[src] beeps: Safety protocols nonexistent!</span>")
```
This needs to use the new procs.

SpaceDragon00 replied:
Sorry, silly mistake on my end

SpaceDragon00 committed:
Using the New Span Procs

- The new visible_message call now uses span_notice() instead of the original span system
